AX: Allow AXCoreObject::supportsPressAction to match the first descendant of a clickable ancestor, even if there are other descendants who would otherwise match

twilco opened the pull request:
#### ddb0937139f40dab201e99762987ac50dbf39c42
<pre>
AX: Allow AXCoreObject::supportsPressAction to match the first descendant of a clickable ancestor, even if there are other descendants who would otherwise match
<a href="https://bugs.webkit.org/show_bug.cgi?id=298662">https://bugs.webkit.org/show_bug.cgi?id=298662</a>
<a href="https://rdar.apple.com/160294153">rdar://160294153</a>

Reviewed by Joshua Hoffman.

Our previous heuristic was AXCoreObject::supportsPressAction was too restrictive, preventing this repair from coming
into affect on a real webpage used for booking a haircut. Relax the heuristic a bit, while still not making it apply
too much so that VoiceOver users don&apos;t hear / read &quot;clickable&quot; all over the place. We balance this by making the heuristic
now require cursor:pointer, as lacking that, the visual experience also doesn&apos;t indicate pressability.

* LayoutTests/accessibility/mac/press-action-not-exposed-for-event-delegation-expected.txt:
* LayoutTests/accessibility/mac/press-action-not-exposed-for-event-delegation.html:
* LayoutTests/accessibility/mac/supports-press-action-with-dynamic-cursor-type-expected.txt: Added.
* LayoutTests/accessibility/mac/supports-press-action-with-dynamic-cursor-type.html: Added.
* LayoutTests/accessibility/onclick-handlers.html:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::supportsPressAction const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::isCursorClickable const):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXNotifications.h:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onStyleChange):
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AXObjectRareData.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::convertToPropertyFlag):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: <a href="https://commits.webkit.org/299991@main">https://commits.webkit.org/299991@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/63a00bddc1d41be20a4e8ebbca9c96064514dacd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120665 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40359 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31011 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127057 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72740 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122541 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41056 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48936 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91649 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60903 "An unexpected error occured. Recent messages:Printed configuration") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123617 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32785 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108161 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72198 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31812 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26268 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70663 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102273 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26450 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129926 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47586 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36128 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47953 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104343 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25452 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45556 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23585 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44246 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47448 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53153 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46917 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50263 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48603 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d70cce748656381cf1edbacb0acdafb31ec9033c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50546

twilco committed:
AX: Allow AXCoreObject::supportsPressAction to match the first descendant of a clickable ancestor, even if there are other descendants who would otherwise match
https://bugs.webkit.org/show_bug.cgi?id=298662
rdar://160294153

Reviewed by Joshua Hoffman.

Our previous heuristic was AXCoreObject::supportsPressAction was too restrictive, preventing this repair from coming
into affect on a real webpage used for booking a haircut. Relax the heuristic a bit, while still not making it apply
too much so that VoiceOver users don't hear / read "clickable" all over the place. We balance this by making the heuristic
now require cursor:pointer, as lacking that, the visual experience also doesn't indicate pressability.

* LayoutTests/accessibility/mac/press-action-not-exposed-for-event-delegation-expected.txt:
* LayoutTests/accessibility/mac/press-action-not-exposed-for-event-delegation.html:
* LayoutTests/accessibility/mac/supports-press-action-with-dynamic-cursor-type-expected.txt: Added.
* LayoutTests/accessibility/mac/supports-press-action-with-dynamic-cursor-type.html: Added.
* LayoutTests/accessibility/onclick-handlers.html:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::supportsPressAction const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::isCursorClickable const):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXNotifications.h:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onStyleChange):
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AXObjectRareData.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::convertToPropertyFlag):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: https://commits.webkit.org/299991@main
